### PR TITLE
thorvg viewer: introduce thorvg viewer

### PR DESCRIPTION
**Pull request by kimcinoo**
Support Emscripten wasm build for thorvg viewer

### CODE REVIEW

File: src/wasm/thorvgwasm.cpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
+#include <thorvg.h>
+
```
Comment (hermet) [inline src/wasm/thorvgwasm.cpp:2]: For less code, 

using namespace std; 
using namespace tvg; 

and remove all other std & tvg prefixes here.

File: src/wasm/thorvgwasm.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        const char *cdata = data.c_str();
+        mErrorMsg = data;
+        if (mPicture->load(cdata, strlen(cdata)) != tvg::Result::Success) {
+            mErrorMsg = "Load failed";
+            return false;
+        }
+
+        getDefaultSize();
+
+        mIsFirstRender = true;
```
Comment (hermet) [inline src/wasm/thorvgwasm.cpp:53]: remove.

File: src/wasm/thorvgwasm.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+            mErrorMsg = "Canvas is NULL";
+            return val(typed_memory_view<uint8_t>(0, nullptr));
+        }
+
+        if (!mPicture) {
+            mErrorMsg = "Picture is NULL";
+            return val(typed_memory_view<uint8_t>(0, nullptr));
+        }
+
+        mWidth = width;
```
Comment (hermet) [inline src/wasm/thorvgwasm.cpp:72]: does it necessary re-allocate memory even though size is not changed?

File: src/wasm/thorvgwasm.cpp [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        mErrorMsg = "None";
+        mIsFirstRender = true;
+
+        tvg::Initializer::init(tvg::CanvasEngine::Sw, 0);
+        mSwCanvas = tvg::SwCanvas::gen();
+        if (!mSwCanvas) {
+            mErrorMsg = "Canvas get failed";
+            return;
+        }
+        mCanvas = mSwCanvas.get();
```
Comment (hermet) [inline src/wasm/thorvgwasm.cpp:124]: any reason to get mCanvas? you can use mSwCanvas directly.

File: src/wasm/thorvgwasm.cpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        float w, h;
+        mPicture->viewbox(nullptr, nullptr, &w, &h);
+        mDefaultWidth = static_cast<uint32_t>(w);
+        mDefaultHeight = static_cast<uint32_t>(h);
+    }
+
+private:
+    std::string mErrorMsg;
+    std::unique_ptr< tvg::SwCanvas > mSwCanvas;
+    tvg::SwCanvas*                   mCanvas;
```
Comment (hermet) [inline src/wasm/thorvgwasm.cpp:152]: looks unnecessary

File: src/wasm/thorvgwasm.cpp [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+        tvg::Initializer::init(tvg::CanvasEngine::Sw, 0);
+        mSwCanvas = tvg::SwCanvas::gen();
+        if (!mSwCanvas) {
+            mErrorMsg = "Canvas get failed";
+            return;
+        }
+        mCanvas = mSwCanvas.get();
+
+        mPicture = tvg::Picture::gen().release();
```
Comment (hermet) [inline src/wasm/thorvgwasm.cpp:126]: I think you can use load() method.

**Review by hermet — CHANGES_REQUESTED** [review]
Can you handle the routine like this?
If possible, That would be cleaner code i think

load()
 1. load file.
 2. picture->scale() (if necessary)
 3. canvas->push(picture)

update(view size):
 1. update canvas buffer (if size is changed)
 2. picture->scale()
 3. canvas->update()

render():
  1. canvas->draw()
  2. canvas->sync()

**Comment by kimcinoo** [comment]
?

File: src/wasm/thorvgwasm.cpp [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+            return;
+        }
+
+        if (!mPicture) {
+            mErrorMsg = "Picture is NULL";
+            return;
+        }
+
+        if (mWidth == width && mHeight == height) {
+            mErrorMsg = "Size is same";
```
Comment (hermet) [inline src/wasm/thorvgwasm.cpp:80]: Does this error?

**Review by hermet — COMMENTED** [review]
a trivial comment.

File: src/wasm/thorvgwasm.cpp [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+    void getDefaultSize()
+    {
+        float w, h;
+        mPicture->viewbox(nullptr, nullptr, &w, &h);
+        mDefaultWidth = static_cast<uint32_t>(w);
+        mDefaultHeight = static_cast<uint32_t>(h);
+    }
+
+    void setScale(int width, int height)
```
Comment (hermet) [inline src/wasm/thorvgwasm.cpp:139]: Does this necessary to realloc though size is same?

void setScale() {
if (size is same) return;
...
}

**Review by hermet — CHANGES_REQUESTED** [review]
Please double-check comments.

File: src/wasm/thorvgwasm.cpp (hunk 7 above)
Comment (kimcinoo) [reply src/wasm/thorvgwasm.cpp]: done

File: src/wasm/thorvgwasm.cpp (hunk 8 above)
Comment (kimcinoo) [reply src/wasm/thorvgwasm.cpp]: done

**Comment by kimcinoo** [comment]
amended based on review opinion.

**Comment by kimcinoo** [comment]
Update Readme.md

**Comment by kimcinoo** [comment]
Remove unintended change

File: src/wasm/thorvgwasm.cpp [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+        mScale = scaleX < scaleY ? scaleX : scaleY;
+
+        mPicture->scale(mScale);
+    }
+
+private:
+    string                 mErrorMsg;
+    unique_ptr< SwCanvas > mSwCanvas;
+    Picture*               mPicture;
+    unique_ptr<uint8_t[]>  mBuffer;
```
Comment (hermet) [inline src/wasm/thorvgwasm.cpp:176]: = nullptr; initialize?

File: src/wasm/thorvgwasm.cpp [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+        float scaleY = static_cast<float>(mHeight) / static_cast<float>(mDefaultHeight);
+        mScale = scaleX < scaleY ? scaleX : scaleY;
+
+        mPicture->scale(mScale);
+    }
+
+private:
+    string                 mErrorMsg;
+    unique_ptr< SwCanvas > mSwCanvas;
+    Picture*               mPicture;
```
Comment (hermet) [inline src/wasm/thorvgwasm.cpp:175]: = nullptr; initialize?

File: src/wasm/thorvgwasm.cpp [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+        float scaleX = static_cast<float>(mWidth) / static_cast<float>(mDefaultWidth);
+        float scaleY = static_cast<float>(mHeight) / static_cast<float>(mDefaultHeight);
+        mScale = scaleX < scaleY ? scaleX : scaleY;
+
+        mPicture->scale(mScale);
+    }
+
+private:
+    string                 mErrorMsg;
+    unique_ptr< SwCanvas > mSwCanvas;
```
Comment (hermet) [inline src/wasm/thorvgwasm.cpp:174]: = nullptr; initialize?

File: src/wasm/thorvgwasm.cpp [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+        mWidth = width;
+        mHeight = height;
+        mBuffer = make_unique<uint8_t[]>(mWidth * mHeight * 4);
+        mSwCanvas->target((uint32_t *)mBuffer.get(), mWidth, mWidth, mHeight, SwCanvas::ABGR8888);
+
+        setScale();
+    }
+
+    void setScale()
+    {
```
Comment (hermet) [inline src/wasm/thorvgwasm.cpp:164]: if (!mPicture) return;

File: src/wasm/thorvgwasm.cpp [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+
+        if (pw == mDefaultWidth && ph == mDefaultHeight) {
+            return;
+        }
+
+        setScale();
+    }
+
+    void setTarget(int width, int height)
+    {
```
Comment (hermet) [inline src/wasm/thorvgwasm.cpp:150]: if (!mSwCanvas) return;

File: src/wasm/thorvgwasm.cpp [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+        pw = mDefaultWidth;
+        ph = mDefaultHeight;
+        mDefaultWidth = static_cast<uint32_t>(w);
+        mDefaultHeight = static_cast<uint32_t>(h);
+
+        if (pw == mDefaultWidth && ph == mDefaultHeight) {
+            return;
+        }
+
+        setScale();
```
Comment (hermet) [inline src/wasm/thorvgwasm.cpp:146]: getDefaultSize() sounds it just gets default size. nothing else.
I think calling setScale()  here is not proper here.

File: src/wasm/thorvgwasm.cpp [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    void setTarget(int width, int height)
```
Comment (hermet) [inline src/wasm/thorvgwasm.cpp:149]: update buffer, update picture scale..
I think updateSize() is better understandable in this case.
setTarget() => updateSize();

**Review by hermet — CHANGES_REQUESTED** [review]
Please check comments.

File: src/wasm/thorvgwasm.cpp (hunk 11 above)
Comment (kimcinoo) [reply src/wasm/thorvgwasm.cpp:174]: done

File: src/wasm/thorvgwasm.cpp (hunk 10 above)
Comment (kimcinoo) [reply src/wasm/thorvgwasm.cpp:175]: done

File: src/wasm/thorvgwasm.cpp (hunk 9 above)
Comment (kimcinoo) [reply src/wasm/thorvgwasm.cpp:176]: done

File: src/wasm/thorvgwasm.cpp (hunk 12 above)
Comment (kimcinoo) [reply src/wasm/thorvgwasm.cpp:164]: done

File: src/wasm/thorvgwasm.cpp (hunk 13 above)
Comment (kimcinoo) [reply src/wasm/thorvgwasm.cpp:150]: done

File: src/wasm/thorvgwasm.cpp (hunk 15 above)
Comment (kimcinoo) [reply src/wasm/thorvgwasm.cpp:149]: done

File: src/wasm/thorvgwasm.cpp (hunk 14 above)
Comment (kimcinoo) [reply src/wasm/thorvgwasm.cpp:146]: done

**Comment by kimcinoo** [comment]
Updated based on review comments

File: src/lib/tvgPictureImpl.h [hunk 16]
```diff
@@ -22,6 +22,7 @@
 #ifndef _TVG_PICTURE_IMPL_H_
 #define _TVG_PICTURE_IMPL_H_
 
+#include <string>
```
Comment (JSUYA) [inline src/lib/tvgPictureImpl.h:25]: Why we need this include?

File: src/loaders/svg/tvgSvgPath.cpp [hunk 17]
```diff
@@ -21,6 +21,8 @@
  */
 #include <string.h>
 #include <math.h>
+#include <ctype.h>
```
Comment (JSUYA) [inline src/loaders/svg/tvgSvgPath.cpp:24]: Why we need this include?

File: src/loaders/svg/tvgSvgSceneBuilder.cpp [hunk 18]
```diff
@@ -20,6 +20,7 @@
  * SOFTWARE.
  */
 #include <math.h>
+#include <string>
```
Comment (JSUYA) [inline src/loaders/svg/tvgSvgSceneBuilder.cpp:23]: Why we need this include?

File: src/lib/tvgPictureImpl.h (hunk 16 above)
Comment (kimcinoo) [reply src/lib/tvgPictureImpl.h:25]: Because of build error.

File: src/loaders/svg/tvgSvgPath.cpp (hunk 17 above)
Comment (kimcinoo) [reply src/loaders/svg/tvgSvgPath.cpp:24]: Because of build error.

File: src/loaders/svg/tvgSvgSceneBuilder.cpp (hunk 18 above)
Comment (kimcinoo) [reply src/loaders/svg/tvgSvgSceneBuilder.cpp:23]: Because of build error.

**Comment by kimcinoo** [comment]
Answer questions

**Review by hermet — APPROVED** [review]

**Comment by hermet** [comment]
@JSUYA more comments?

**Review by JSUYA — APPROVED** [review]
no more comment

**Pull request merged**